bpf: fib: always use bpf_redirect_neigh() when available

jrife opened the pull request:
In some environments, neighbor entries for non-cluster hosts remain stale forever, breaking connectivity in cases where their IP is reassigned to a new host with a different MAC address. This isn't a problem for hosts within the cluster; Cilium sets the `managed` and `extern_learn` flags for those neighbor entries, so the kernel actively refreshes them.

In our particular case, we have a set of VMs forming our Kubernetes cluster running a NodePort service and another VM in the same L2 domain talking to that service. This set of VMs is upgraded as a unit by spinning up new VMs, draining the old ones, and reassigning the IP addresses, although not necessarily in that order. At some point during this process, the MAC address assigned to the IP for that external VM changes. The network switches in our setup enable ARP suppression which, as I understand it, suppresses GARPs if the ARP suppression cache on the switch already has an entry for a given MAC address. This, combined with Cilium's current behavior leads to a situation where we we need to apply some workarounds to get things talking again. What we hope would occur is that cluster nodes would send an ARP request for any STALE neighbor entries in their ARP table, but currently this never happens.

`fib_redirect()` and others call `bpf_fib_lookup()` and pass its result along to `fib_do_redirect()`. Currently, `fib_do_redirect()` only uses `bpf_redirect_neigh()` if the FIB lookup was unsuccessful, returning `BPF_FIB_LKUP_RET_NO_NEIGH`. Otherwise, it "manually" populates the L2 header of the packet and uses `bpf_redirect()` to send it out through the host interface. However, `bpf_fib_lookup()` returns `BPF_FIB_LKUP_RET_SUCCESS` regardless of the state of the neighbor entry, and `bpf_redirect()` doesn't trigger a neighbor event in the kernel. The net result is that stale entries can remain stale which is problematic in setups like the one described above.

* The first commit makes it so Cilium always uses `bpf_redirect_neigh()` when available to trigger refresh of stale entries and avoid corner cases like ours where invalid entries can can remain stale forever. 

* The last two commits use the `BPF_FIB_LOOKUP_SKIP_NEIGH` [1] flag in preceding calls to `bpf_fib_lookup()` to avoid duplicate work in cases where we want to defer to `bpf_redirect_neigh()` for the neighbor resolution. The first commit introduces a probe that ensures that `BPF_FIB_LOOKUP_SKIP_NEIGH` is supported on the current system, as some older kernels may support `bpf_fib_lookup()` but not the `BPF_FIB_LOOKUP_SKIP_NEIGH` flag. The second commit plumbs through the probe result to let the FIB code decide whether or not to set it.

[1]: [`bpf: Add BPF_FIB_LOOKUP_SKIP_NEIGH for bpf_fib_lookup`](https://github.com/torvalds/linux/commit/31de4105f00d64570139bc5494a201b0bd57349f)

### Without This Change

The neighbor entry for 192.168.4.255, a VM in the same L2 domain as my cluster nodes, remains STALE after making an HTTP request to a NodePort service hosted on that node.

```
jordan@crow:~$ ip neigh | grep 225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE
...
jordan@crow:~$ ip neigh | grep 225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE
```

### With This Change

The HTTP request from 192.168.4.255 triggers a neighbor event on the cluster node, so the neighbor entry transitions from STALE to REACHABLE.

```
jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE
...
jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 DELAY
...
jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 REACHABLE
```

Fixes: #34503

```release-note
bpf: fib: Fix issue where neighbor entries remain stale forever in some cases.
```

ysksuzuki reviewed (commented):
Thank you for the PR! I have two questions.

- Have you considered making this behavior, where bpf_redirect_neigh() is always used when available, opt-in instead of enabling it by default? Or have you already evaluated that the potential performance impact would be trivial?
- Does this issue occur only when the destination is a host outside the cluster? If so, would it be worth considering an optimization where bpf_redirect_neigh() is always used only when the destination is outside the cluster?

julianwiedmann commented on bpf/lib/egress_gateway.h:
```diff
@@ -43,7 +43,9 @@ int egress_gw_fib_lookup_and_redirect(struct __ctx_buff *ctx, __be32 egress_ip,
-	ret = (__s8)fib_lookup_v4(ctx, &fib_params, egress_ip, daddr, 0);
+	ret = (__s8)fib_lookup_v4(ctx, &fib_params, egress_ip, daddr,
+				  neigh_resolver_available()
+				  ? BPF_FIB_LOOKUP_SKIP_NEIGH : 0);
```
Just to reduce the churn in the patch, you could push this logic down into the `fib_lookup_v4()` and `fib_lookup_v6()` helpers.

jrife replied:
> Just to reduce the churn in the patch, you could push this logic down into the fib_lookup_v4() and fib_lookup_v6() helpers.

IIRC there was some reason I didn't do this, but I'll take another look and see if it's possible.

jrife replied:
I went ahead and pushed this into `fib_lookup_v(4|6)`. My main concern before was that setting the flag inside the function doesn't make sense if there are callers that don't immediately follow up with a call to `fib_do_redirect`. However, as far as I can tell, this seems to be the usage pattern all over. Feel free to take another look.

*Note*: Unit tests are passing, but looks like some CI issue is currently blocking image builds for smoke tests, conformance tests, etc. Filed https://github.com/cilium/cilium/issues/39600

jrife committed:
bpf: fib: Add probe for BPF_FIB_LOOKUP_SKIP_NEIGH support

Prepare for the next patch which passes BPF_FIB_LOOKUP_SKIP_NEIGH to
bpf_fib_lookup in cases where we want to defer to bpf_redirect_neigh for
neighbor resolution by adding a probe to see if the current system
supports it.

Fixes: cilium#34503
Signed-off-by: Jordan Rife <jrife@google.com>

jrife committed:
bpf: fib: Use BPF_FIB_LOOKUP_SKIP_NEIGH when available

Use BPF_FIB_LOOKUP_SKIP_NEIGH in cases where the system supports it and
we want to defer to bpf_redirect_neigh for neighbor resolution.

Fixes: cilium#34503
Signed-off-by: Jordan Rife <jrife@google.com>

jrife committed:
bpf: fib: always use bpf_redirect_neigh() when available

In some environments, neighbor entries for non-cluster hosts remain
stale forever, breaking connectivity in cases where their IP is
reassigned to a new host with a different MAC address. This isn't a
problem for hosts within the cluster; Cilium sets the `managed` and
`extern_learn` flags for those neighbor entries, so the kernel actively
refreshes them.

In our particular case, we have a set of VMs forming our Kubernetes
cluster running a NodePort service and another VM in the same L2
domain talking to that service. This set of VMs is upgraded as a unit
by spinning up new VMs, draining the old ones, and reassigning the IP
addresses, although not necessarily in that order. At some point during
this process, the MAC address assigned to the IP for that external VM
changes. The network switches in our setup enable ARP suppression
which, as I understand it, suppresses GARPs if the ARP suppression cache
on the switch already has an entry for a given MAC address. This,
combined with Cilium's current behavior leads to a situation where we
we need to apply some workarounds to get things talking again. What we
hope would occur is that cluster nodes would send an ARP request for
any STALE neighbor entries in their ARP table, but currently this never
happens.

fib_redirect() and others call bpf_fib_lookup() and pass its result
along to fib_do_redirect(). Currently, fib_do_redirect() only uses
bpf_redirect_neigh() if the FIB lookup was unsuccessful, returning
BPF_FIB_LKUP_RET_NO_NEIGH. Otherwise, it "manually" populates the L2
header of the packet and uses bpf_redirect() to send it out through the
host interface. However, bpf_fib_lookup() returns
BPF_FIB_LKUP_RET_SUCCESS regardless of the state of the neighbor entry,
and bpf_redirect() doesn't trigger a neighbor event in the kernel. The
net result is that stale entries can remain stale which is problematic
in setups like the one described above.

Always use bpf_redirect_neigh() if available to trigger refresh of stale
entries and avoid corner cases like ours where invalid entries can
can remain stale forever.

Without This Change
-------------------
The neighbor entry for 192.168.4.255, a VM in the same L2 domain as my
cluster nodes, remains STALE after making an HTTP request to a NodePort
service hosted on that node.

jordan@crow:~$ ip neigh | grep 225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE
...
jordan@crow:~$ ip neigh | grep 225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE

With This Change
----------------
The HTTP request from 192.168.4.255 triggers a neighbor event on the
cluster node, so the neighbor entry transitions from STALE to
REACHABLE.

jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 STALE
...
jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 DELAY
...
jordan@crow:~$ ip neigh | grep 192.168.4.225
192.168.4.225 dev enp5s0 lladdr bc:24:11:93:7f:d4 REACHABLE

Fixes: cilium#34503
Signed-off-by: Jordan Rife <jrife@google.com>

borkmann commented on bpf/lib/fib.h:
```diff
@@ -91,61 +91,65 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+						   sizeof(nh_params), 0);
+		}
+
+		return (int)redirect_neigh(oif, NULL, 0, 0);
```
small nit here and elsewhere: do we need the explicit `(int)` casts? I'd remove them if not since the function already returns int in its signature. rest lgtm, lets get this in! thanks Jordan!

jrife replied:
Maybe not, but this is how they were before; I just removed lines and changed indention in this PR. It might be better to open another PR for `(int)` cleanup here and other places.

pchaigno replied:
Yep, looks like we're already doing this in multiple other places, so not going to block on it. Though if we're touching this code and you're repushing anyway, it's probably worth fixing.

jrife replied:
```
/go/src/github.com/cilium/cilium/bpf/lib/fib.h:111:10: error: implicit conversion loses integer precision: 'long' to 'int' [-Werror,-Wshorten-64-to-32]
  111 |                 return redirect_neigh(oif, NULL, 0, 0);
      |                 ~~~~~~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

It looks like the casts are necessary to make the compiler happy. Probably why they were there before.

rgo3 commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -435,6 +436,41 @@ func HaveIPv6Support() error {
+		Data:   make([]byte, 14),
+		Repeat: 1,
+	})
+	var ve *ebpf.VerifierError
```
This doesn't look right to me. Shouldn't the check for a verifier error happen after loading the program, in this case `LoadProbesObjects`?

jrife replied:
Yes. Good catch. I moved some things around and forgot to adjust where I do the `VerifierError` check. Thanks!

rgo3 commented on pkg/datapath/linux/probes/probes.go:
```diff
@@ -435,6 +436,41 @@ func HaveIPv6Support() error {
+		}
+	}
+	if err != nil {
+		return fmt.Errorf("loading eBPF collection into the kernel: %w", err)
```
This should indicate that the error happened when trying to run the program, as loading has already succeeded previously
```suggestion
		return fmt.Errorf("running probe program: %w", err)
```

rgo3 commented on pkg/datapath/linux/requirements.go:
```diff
@@ -114,6 +114,18 @@ func CheckRequirements(log *slog.Logger) error {
+					logfields.Error, err,
+				)
+			} else {
+				log.Warn("Unable to determine if BPF_FIB_LOOKUP_SKIP_NEIGH is supported; it will not be used",
```
What is the expected outcome of this probe on kernels <6.6? Does the probe return `ErrNotSupported` or another error code?
I'm asking because for most of the other probes, unless the error is `ErrNotSupported` we return an error and not just warn users about something that isn't expected to happen.

jrife replied:
> What is the expected outcome of this probe on kernels <6.6?

It should return `ErrNotSupported` if the flag isn't available on the current kernel. I confirmed this by observing CI and seeing runs on older kernels spit out that message.

> I'm asking because for most of the other probes, unless the error is ErrNotSupported we return an error and not just warn users about something that isn't expected to happen.

I considered returning an error instead, but didn't want to block startup on this requirement, since it's not strictly necessary for correct datapath operation (it's more of an optimization).

rgo3 replied:
>I considered returning an error instead, but didn't want to block startup on this requirement, since it's not strictly necessary for correct datapath operation (it's more of an optimization).

This is a valid concern, but I'd like to maintain consistency across all probes we implement for Cilium. For Cilium, we've decided that probes should return conclusive results: `nil` or `ErrNotSupported`. Anything else is unexpected and shouldn't happen under normal operation, so it likely hints at a bigger issue and warrants a failure. See `HaveProgramHelper` as an example, only conclusive results are returned to callers, anything else fatals.

Unless this probe depends on specific kernel configuration that isn't required to run Cilium by default, I'd prefer to keep the same semantics and error handling across the package.

jrife replied:
> we've decided that probes should return conclusive results: nil or ErrNotSupported

Alright, I've done my best to adhere to this with the introduction of `HaveFibLookupSkipNeighNoFail` which guarantees that only `ErrNotSupported` or `nil` is returned much like some of the other probes. However, I've kept the original `HaveFibLookupSkipNeigh` as well. The relationship between `HaveFibLookupSkipNeighNoFail` and `HaveFibLookupSkipNeigh` mirrors that of, e.g., `probes.HaveProgramHelper` and `features.HaveProgramHelper` where the "outer" function guarantees only `ErrNotSupported` or `nil` are returned.

I considered just having a single `HaveFibLookupSkipNeigh` that does it all, but since it would need to accept a `logger` it didn't really make sense in the `pkg/datapath/loader/config.go` context:

```
node.SupportsFibLookupSkipNeigh = probes.HaveFibLookupSkipNeigh() == nil
```

I'd either have to pass in `nil` or plumb through a logger that isn't used. This seemed awkward, so I kept `HaveFibLookupSkipNeigh` for use inside `nodeConfig()`.

rgo3 replied:
I probably chose the wrong wording here, I would have been fine with just doing the following in `requirements.go` which would eliminate the need for the wrapper. But consider this as a nit:

```golang
if err := probes.HaveFibLookupSkipNeigh(); err != nil {
    if !errors.Is(err, probes.ErrNotSupported) {
	    return errors.New("Unable to determine if BPF_FIB_LOOKUP_SKIP_NEIGH is supported")
    }
    log.Info("BPF_FIB_LOOKUP_SKIP_NEIGH is not supported; it will not be used", logfields.Error, err)
}
```

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -195,7 +195,9 @@ nodeport_fib_lookup_and_redirect(struct __ctx_buff *ctx,
-	ret = (int)fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l), 0);
+	ret = (int)fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l),
+			      neigh_resolver_available()
+			      ? fib_lookup_skip_neigh() : 0);
```
Nit (i.e., not going to block on this): It feels like this would be a bit more readable with a new helper function:
```c
static __always_inline __u64 fib_lookup_neigh_flags(void)
{
        return neigh_resolver_available()? fib_lookup_skip_neigh() : 0;
}
```
which you could then just use in the 4 places to mask the condition. Or you could also fold the condition in `fib_lookup_skip_neigh()` directly.

jrife replied:
Done. I just rolled this all into `fib_lookup_skip_neigh` to make things simpler.

pchaigno reviewed (approved):
LGTM assuming Robin's comments are addressed.

rgo3 reviewed (approved):
Only left one more nit, otherwise this LGTM for loader changes. Thanks for your patience Jordan!

julianwiedmann commented on bpf/include/bpf/config/node.h:
```diff
@@ -24,3 +24,6 @@ NODE_CONFIG(__u32, trace_payload_len, "Length of payload to capture when tracing
 
 NODE_CONFIG(__u32, direct_routing_dev_ifindex, "Index of the interface used to connect nodes in the cluster.")
+
+NODE_CONFIG(bool, supports_fib_lookup_skip_neigh,
+	    "Whether or not BPF_FIB_LOOKUP_SKIP_NEIGH is supported.")
```
For my own education - how would we accurately test this in `bpf/complexity-tests` ?

In the past we would simply add the relevant `HAVE_FEATURE_FOO` macro to the configs for relevant kernel versions. But for load-time configs that's no longer possible ...?

pchaigno replied:
See https://github.com/cilium/cilium/pull/40367.

julianwiedmann replied:
:+1: ok so on a casual look the new verifier-loader still needs to learn (for instance [here](https://github.com/cilium/cilium/blob/ce1f55100f689f6d4f24801f90c86578b8672a08/pkg/datapath/loader/verifier_load_test.go#L27)) how to set up `SupportsFibLookupSkipNeigh` according to the tested kernel version.

And that matches the fact that `ci-verifier` currently runs all its permutations with `SupportsFibLookupSkipNeigh:false`, even on `bpf-next`.

dylandreimerink replied:
Yes, when we add new load time config option or convert from compile time we should add them to the ...Permutations function(s). We still have to figure out which combinations we want to run and which we don't care about as much (we are likely most interested in combinations that result in higher complexity, not all permutations possible).

Since that logic in Go, we should indeed be able to use probes instead of hard-coding settings to certain kernel versions.
